python/lsst/daf/butler/core/logging.py

```diff
@@ -21,16 +21,63 @@
 
 from __future__ import annotations
 
-__all__ = ("VERBOSE", "ButlerMDC")
+__all__ = ("VERBOSE", "ButlerMDC", "ButlerLogRecords", "ButlerLogRecordHandler",
+           "ButlerLogRecord", "JsonFormatter")
 
 import logging
+import datetime
+import traceback
+from typing import List, Union, Optional, ClassVar, Iterable, Iterator, Dict, IO
+
+from logging import LogRecord, StreamHandler, Formatter
+from pydantic import BaseModel
 
 VERBOSE = (logging.INFO + logging.DEBUG) // 2
 """Verbose log level"""
 
+_LONG_LOG_FORMAT = "{levelname} {asctime} {name} {filename}:{lineno} - {message}"
+"""Default format for log records."""
+
 logging.addLevelName(VERBOSE, "VERBOSE")
 
 
+def _isplit(string: Union[bytes, str], sep: Optional[Union[bytes, str]]) -> Iterator[Union[bytes, str]]:
```

Comment:
@andy-slac This is what I came up with for a split generator. I initially wrote:

```diff
diff --git a/python/lsst/daf/butler/core/logging.py b/python/lsst/daf/butler/core/logging.py
index 2a6ac1ab..ede1e6fe 100644
--- a/python/lsst/daf/butler/core/logging.py
+++ b/python/lsst/daf/butler/core/logging.py
@@ -41,6 +41,30 @@ _LONG_LOG_FORMAT = "{levelname} {asctime} {name} {filename}:{lineno} - {message}
 logging.addLevelName(VERBOSE, "VERBOSE")
 
 
+def _isplit(string: Union[bytes, str], sep: Optional[Union[bytes, str]]) -> Iterator[Union[bytes, str]]:
+    """Generator version of split to avoid making potentially large list.
+
+    Notes
+    -----
+    Python `str` do not support `memoryview` so slicing the string on each
+    iteration to chop the front off will copy the entire remainder of the
+    string. This implementation therefore keeps track of the start and end
+    positions.
+    """
+    if sep is None:
+        sep = " " if isinstance(string, str) else b" "
+    end = None
+    while end != -1:
+        # the type signature above does not explain that the string and
+        # sep types must match up.
+        end = string.find(sep)  # type: ignore
+        if end != -1:
+            yield string[:end]
+            string = string[end + 1:]
+    yield string
+    return
+
```
but then learned that `string = string[end:]` generates a copy so you would still end up with double the memory being required if the string is large.

mypy hates the current implementation and I'm not entirely sure what I have to do to tell mypy:

* All the parameters given and returned are the same type and not arbitrary combinations.
* That the result of `string[0]` is an `int` if `string` is `bytes` -- mypy really gets upset about the line that converts the int back to bytes again.

Reply:
What about something like 

    def _isplit(string, sep):
        begin = 0
        while True:
            end = string.find(sep, begin)
            yield string[begin:end]
            if end == -1:
                return
            begin = end + 1

Reply:
Ha. Serves me write for not reading the `str.find()` docs completely and so not noticing the `begin` parameter. Thanks.